PLANET-4603 Introduce campaign sidebar

Inwerpsel opened the pull request:
This requires the following changes in master-theme: https://github.com/greenpeace/planet4-master-theme/pull/984

* Register a sidebar with campaign settings.
* Add HOC function for creating post-meta controls.
* Use new meta field without underscore for campaign_page_template, as
the underscore prevents API usage. In usages fall back to old field if
the new doesn't exist yet.
* Add ColorPaletteControl that wraps around ColorPalette and functions
similarly to other input controls. This is currently missing in
wordpress, but should it be added there then this can go.

pablocubico commented:
Hey @Inwerpsel this looks great!

I accidentally clicked "Unpin from toolbar" and the sidebar icon disappeared, apparently you have to register a "MoreMenuItem" so it appears in the three-dots "More" menu in the toolbar and you can "pin" it again. Could you add that? I found an example here: https://github.com/WordPress/gutenberg/issues/14211

I would suggest two design tweaks:
1 - just adding 15px padding to the `.components-base-control` wrapper in the template selector, so it matches the rest of the controls there, maybe you can add a specific class name to that element to avoid overriding other controls using that class.
2 - the radio buttons are misaligned, we are probably overriding those elsewhere

Looks amazing! Can't wait to do more changes in this direction.

Inwerpsel commented:
> Hey @Inwerpsel this looks great!
> 
> I accidentally clicked "Unpin from toolbar" and the sidebar icon disappeared, apparently you have to register a "MoreMenuItem" so it appears in the three-dots "More" menu in the toolbar and you can "pin" it again. Could you add that? I found an example here: [WordPress/gutenberg#14211](https://github.com/WordPress/gutenberg/issues/14211)
> 
> I would suggest two design tweaks:
> 1 - just adding 15px padding to the `.components-base-control` wrapper in the template selector, so it matches the rest of the controls there, maybe you can add a specific class name to that element to avoid overriding other controls using that class.
> 2 - the radio buttons are misaligned, we are probably overriding those elsewhere
> 
> Looks amazing! Can't wait to do more changes in this direction.

Thanks for the feeback @pablocubico ! Maybe you want to check again as I removed the options from withPostMeta and moved them to a separate function that gets these from a template file instead of from an endpoint.

I added the menu link you proposed, indeed we need to have it or there is no way for a user to get the menu back after unpinning. The `PluginSidebar` component also has `isPinned` and `isPinnable` props, unfortunately it is not possible to to set `isPinned` to true and `isPinnable` to false (probably they want to avoid plugins cluttering the editor).

For the 15px margin, does that still apply? Currently it looks like it matches the rest. Maybe your comment is on an earlier version. I had added the `<div class="components-panel__body is-opened">` there manually as the defaults styles of the sidebar only work when putting things in a collapsible panel, but it's a bit tedious for the end user to have everything inside a panel. But maybe we need to write some css so that it always looks good when you include controls outside a panel without having to provide a wrapper diff.

The aligment of the button and the text for the radio buttons looks like a general issue. For example in renderEdit of the SocialMedia block it is also off by the same amount.

pablocubico commented on assets/src/components/Sidebar/CampaignSidebar.js:
```diff
@@ -0,0 +1,225 @@
+        <PluginSidebarMoreMenuItem
+          target={ CampaignSidebar.getId() }
+          icon={ CampaignSidebar.getIcon()}>
+          Your Sidebar Plugin
```
I guess this could be "Campaign Options", same as below.

Inwerpsel replied:
Maybe I'll just remove `PluginSidebarMoreMenuItem`, as I added a css fix that removes the unpin button. So if we remove this one too then the user can't unpin it in the first place.

Inwerpsel replied:
In the end I added it again, just in case the css fix would fail as suggested by Pablo

pablocubico commented on assets/src/components/Sidebar/CampaignSidebar.js:
```diff
@@ -0,0 +1,220 @@
+import { PluginSidebar, PluginSidebarMoreMenuItem } from "@wordpress/edit-post";
```
Hey @Inwerpsel this "PluginSidebarMoreMenuItem" is no longer used, BUT, I missed your comment about this, I guess you may think about re-adding it just in case, since the CSS fix for "Unpin from toolbar" could get translated if you set the WP admin to other languages, also it might be a bit too generic (what if other buttons on other toolbars have that label?). I think you can keep both fixes, adding a bit more scope to the CSS thing.

Inwerpsel replied:
Re-added it :+1:
